src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/JsonObjectDefaultConverter.cs

```diff
@@ -0,0 +1,415 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+
+namespace System.Text.Json.Serialization.Converters
+{
+    /// <summary>
+    /// Default base class implementation of <cref>JsonObjectConverter{T}</cref>.
+    /// </summary>
+    internal sealed class JsonObjectDefaultConverter<T> : JsonObjectConverter<T>
+    {
+        internal override bool OnTryRead(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options, ref ReadStack state, out T value)
+        {
+            bool shouldReadPreservedReferences = options.ReferenceHandling.ShouldReadPreservedReferences();
+            object obj;
+
+            if (!state.SupportContinuation && !shouldReadPreservedReferences)
+            {
+                // Fast path that avoids maintaining state variables and dealing with preserved references.
```

Comment:
Somewhere in the fast path you are missing an `EndProperty()` call which is causing that the json path differs from the slow path.
```cs
private class Node
{
    public Node Child { get; set; }
    public object Sibling { get; set; }
}

[Theory]
// Missing EndObject token.
[InlineData(@"{""Child"":{""$id"":""1""},""Sibling"":{""$ref"":0}")]
[InlineData(@"{""Child"":{""$id"":""1""},""Sibling"":{""$ref"":}}")]
public static void VerifyThrowOnJsonObjectConverter(string json)
{
    JsonException ex = Assert.Throws<JsonException>(() => JsonSerializer.Deserialize<Node>(json, s_deserializerOptionsPreserve));
    JsonException ex2 = Assert.Throws<JsonException>(() => JsonSerializer.Deserialize<Node>(json));

    Assert.Equal(ex2.Message, ex.Message);
    Assert.Equal(ex2.Path, ex.Path);
}
```

Reply:
I believe this is because "$ref" is not a real property so we just skip the rest.

Reply:
Yes, that seems like it. Thanks for clarify.

Reply:
Not really, this seems to happen when there is a missing `EndObject` token.
```cs
[InlineData(@"{""Child"":{""Name"":""foo""},""Sibling"":{""Name"":""bar""}")]
```

However maybe we can fix this on a separate issue that tackles JSON path consistency.

Reply:
I see. Yes, the fast path should have an EndProperty(). I'll add it.